### ISSUE TITLE
SW-01: Spike: hold-detection corpus, model, and on-device feasibility

### ISSUE THREAD

**Issue by marcodejongh** [marcodejongh/boardsesh#5434]
Epic #5346 (SW-01). First issue of the spray-walls epic; it de-risks hold detection before any product code is written.

## Why

The product promise is "photograph your wall, we find the holds". The owner chose **on-device** detection for economy, gated on measured quality. Nothing in the repo does image segmentation of photos today (`packages/moonboard-ocr` reads app screenshots; `scripts/generate-board-art-geometry.ts` traces rendered board art), and the mobile app has no ML runtime. This spike answers three questions with numbers: how well a phone-sized model finds holds on real spray walls, which React Native runtime actually builds on Expo SDK 57 / RN 0.86, and which model family we may legally ship.

## Scope

- `ml/holds/` beside `ml/climb2vec/` (same offline-only Python layout): `README.md` corpus protocol, `label/` (COCO export), `train.py` (one pipeline), `export.py` (ONNX + TFLite fp16/int8; CoreML optional), `eval.py`.
- Corpus: at least 20 real spray-wall photos sourced from Discord users, portrait and landscape, mixed lighting, at least 3 hard cases (wood-on-wood holds, dense sections, volumes, tape). Collect through a `POST /api/spray-wall-test-data` handler cloned from `packages/backend/src/handlers/ocr-test-data.ts` into the `private` bucket. Hand-label every hold. Hold out 5 photos by photo, not by hold.
- `eval.py` scores the **exported artifacts**, never the PyTorch weights, at three configs: nano tiled at 1024 px, small tiled at 1024 px, medium untiled at 1280 px. Metrics: precision / recall / F1 at mask IoU 0.5 and box IoU 0.5 on the held-out split, per-photo runtime, and a 5-photo count of manual corrections a user would need.
- Native mini-spike: which runtime builds with `expo prebuild` + a dev client on SDK 57 / RN 0.86 new architecture and loads a model from a **file path** (not a bundled asset). Try in order: `react-native-fast-tflite`, `react-native-executorch`, `onnxruntime-react-native`. Record device timing and peak memory on a Pixel 6a-class Android and an iPhone 12 for the nano and small models with 2×2 tiling.
- Licence (owner decision 2026-09-14): **Apache-2.0 model families only** in this repo and in anything that ships in the app (RF-DETR, YOLOX, D-FINE, RT-DETR). Ultralytics code and weights (AGPL-3.0) are off the table here; they are allowed only inside the separate AGPL detector microservice repo if the contingency (#5451) triggers. Box-only detectors are acceptable candidates: outlines are cosmetic in the renderer (ring fallback), so evaluate box + a classical in-box segmentation as one config. Record the licence of every Roboflow Universe dataset used; exclude CC BY-NC data; CC BY needs attribution in the app licences screen.
- Commit the fixtures (a few photos with consent, the exported nano ONNX, expected detections) for the Node parity tests in SW-06.

## Out of scope

Any product UI, the runtime PR itself (SW-02), a server container (SW-18), training-data tooling beyond what the corpus needs.

## Acceptance

- A report comment on #5346 with the metric table for the three configs, device timings, the runtime that built, and the licence findings.
- **Acceptance gate (rewritten 2026-09-15 after the AGPL reference benchmark, F1 0.62 on unseen walls):** the old F1 gates are retired. Report, for every config, weighted corrections per hold = (2·miss + 1·false positive) ÷ holds on the spray `eval` split with the threshold swept on `tune` for THAT metric (report the F1-optimal threshold separately), and gesture savings = 1 − (weighted corrections ÷ 2). **Suggestions ship when gesture savings ≥ 40 %.** On-device qualifies if a config clears that within < 500 MB peak memory and a few seconds per photo on the phone; otherwise the same model serves from a CPU container (#5451, rescoped). Keep the AGPL reference (jwlarocque/yolov8n-freeclimbs-detect-2 @ 2560: F1 0.616, ~1.1 weighted corrections, ~45 % savings) in a separate reference table, never in the harness.
- Meets F1 but not speed or memory → recommend SW-18 (server contingency) and say why.
- **Milestone 2026-09-15 (M5 Max full run):** medium-untiled-1280 int8, threshold 0.60 from `tune`: spray-eval F1 0.650 (P 0.695 / R 0.610), corrections 0.66, weighted 1.05 → 47 % gesture savings, gate cleared; AGPL reference 0.616 / 1.14; in-domain (Roboflow test, 8,030 holds) F1 0.906; latency 0.148 s CPU; int8 31.3 MB; curve flat from epoch 6. nano-untiled-1024 in progress; its number decides on-device (medium peaks at 545 MB, over the phone budget).
- Follow-up (not this run): licence-check the Roboflow Universe "Hold Detector" (Climb AI) instance-segmentation set as a mask-training corpus so `eval.py` mask IoU has ground truth; a second quarantined reference (climbingcrux_model, YOLOv9, GPL-3.0 training code) may be scored the same way as the AGPL one if it does not slow training.
- Neither config reaches the gesture-savings gate → recommend manual-first: the hold editor (SW-08) ships alone with a "suggest holds" button running whatever model exists; versioning and matching are unaffected.
- `ml/holds/README.md` documents how to reproduce every number.

## Where

`ml/holds/` (new), `packages/backend/src/handlers/ocr-test-data.ts` (clone), `packages/backend/src/server.ts` (route), `packages/mobile/package.json` (spike branch only, not merged).

## Blocked by / Blocks

- Blocked by: none.
- Blocks: SW-02 (#5435), SW-06 (#5439), SW-18 (#5451).

## Size

L. Python + one native spike branch. Pair with a reviewer who re-runs `eval.py` on the held-out split.

**Comment by marcodejongh** [comment]
## Full training run on the M5 Max (SW-01) — results

Ran the full-dataset recipe from the plan: all 3,876 CC BY train-split photos, 10 epochs, `--device mps`, on the M5 Max / 128 GB. Draft PR with every eval JSON and the README update: #5472.

Protocol: int8 ONNX is what gets scored (as always); the score threshold is swept on the spray-wall `tune` half **against the int8 artifact** (the calibration issue Codex flagged on #5463) and F1 is reported once on the untouched `eval` half — the same 10 photos / 964 holds as every previous number. The spray corpus here is the same tune/eval split as the spike (pulled from the spike box).

### Spray-wall eval — the number that counts

| | nano-untiled **full** | medium-untiled **full** | nano-tiled CC BY (prev) | medium CC BY (prev) | Gate |
| --- | --- | --- | --- | --- | --- |
| Trained on | 3,876 photos × 10 ep | 3,876 photos × 10 ep | 1,400 tiles × 1 ep | 600 photos × 1 ep | — |
| Threshold (tune, int8) | 0.60 | 0.60 | 0.05 | 0.20 | — |
| Precision | 0.682 | 0.695 | 0.228 | 0.514 | — |
| Recall | 0.637 | 0.610 | 0.419 | 0.612 | — |
| **F1 @ IoU 0.5** | **0.659** | 0.650 | 0.295 | 0.559 | ≥ 0.80 ❌ / ≥ 0.65 auto-detect: nano ✅ medium ✅ (barely) |
| Corrections ÷ holds | 0.660 | 0.658 | 2.00 | 0.97 | ≤ 0.10 ❌ |
| Weighted corr. (2·miss+FP) ÷ holds | **1.02** | 1.05 | — | — | ~0.5 (proposed tap budget) ❌ |
| Latency p50 / p95 (M5 Max CPU) | 0.166 / 0.187 s | 0.148 / 0.241 s | 0.59 / 0.66 s¹ | 0.35 / 0.38 s¹ | ≤ 8 s ✅ |
| Peak RSS (onnxruntime CPU) | **987 MB** | 593 MB | 391 MB¹ | 545 MB¹ | < 500 MB ❌ |
| int8 size | 31.3 MB | 32.8 MB | 28.7 MB | 32.8 MB | ~25 MB budget ⚠️ |
| In-domain (Roboflow test, 8,030 holds) F1 | 0.908 | 0.893 | — | — | — |
| Train wall clock | 99 min (MPS) | 78 min (MPS)² | 39 min (CPU) | 34 min (CPU) | — |

¹ spike-box numbers; latency not directly comparable. ² plus one 70-min epoch lost to an MPS allocator stall before the batch/watermark fix; details in the README's training notes.

`nano-tiled-1024` was not rerun: the spike showed 2×2 tiling hurts at this photo scale, and the new full-frame config confirms it from the other side. `nano-untiled-1024` trains at 768 px (not 1024): DINOv2 attention is quadratic in tokens and 1024-px training ran ~70× slower per step on MPS.

### What the run settles

1. **Full-frame nano beats medium** (0.659 vs 0.650) at 31.3 MB. The phone-sized architecture is not the bottleneck.
2. **The dataset is exhausted.** Per-epoch spray scoring (curve files in the PR) plateaus by epoch 4–6 on both configs — medium 0.586 → 0.619, nano 0.590 → 0.615 (fp32 coarse sweep) — while in-domain F1 reaches 0.89–0.91. Everything still missing is **domain shift to real spray walls**. More epochs and bigger models are dead ends; spray-wall training photos (the `POST /api/spray-wall-test-data` corpus) are the only lever left.
3. **Calibration is fixed by full training**: tune-optimal threshold moved 0.05–0.20 → 0.60 on both configs, and the F1-optimal and tap-optimal thresholds now coincide.
4. **The gates measure the wrong thing** (per the discussion today): no model of any licence reaches F1 0.80 on unseen spray walls. On the proposed tap budget (weighted corrections ≤ ~0.5/hold), nobody is there either — best is 1.02 — but in-domain the same model scores **0.33**, i.e. the budget is achievable once training data looks like the target walls.
5. **New on-device blocker: memory.** One 768 px full-frame pass peaks at 987 MB on onnxruntime CPU — double the 500 MB gate. Options for SW-02: tiled *inference* over full-frame-trained weights, smaller eval-time input, or CoreML/NNAPI memory behaviour (all unmeasured).

### AGPL reference (quarantined benchmark, not in the repo)

For calibration of expectations only: `jwlarocque/yolov8n-freeclimbs-detect-2` (YOLOv8n, AGPL-3.0, trained at 2560 px specifically on spray walls, claims 0.988 mAP50 on its own test set), scored in a throwaway venv on the identical tune/eval protocol:

| | AGPL freeclimbs @ 2560 | Our nano-untiled (Apache) |
| --- | --- | --- |
| F1 on spray `eval` | 0.616 | **0.659** |
| Precision / Recall | 0.692 / 0.556 | 0.682 / 0.637 |
| Weighted corr. ÷ holds | 1.14 | **1.02** |

**Our Apache model now beats the best available purpose-built AGPL spray-wall detector.** There is no remaining accuracy case for an AGPL microservice (#5451); if a server-side pass is wanted, our own weights are the engine. Its training set ("freeclimbs v2") is unreleased/unlicensed — nothing to harvest.

### Artifacts

int8 ONNX files are local on the M5 Max (over the repo's 15 MB ceiling, so not in git), heading to R2 under `models/hold-detector/<version>/` as the next step:

- `ml/holds/.data/artifacts/nano-untiled-1024/model-int8.onnx` — 31.3 MB
- `ml/holds/.data/artifacts/medium-untiled-1280/model-int8.onnx` — 32.8 MB

Total GPU time for everything above: ~4.5 h wall clock, one evening on one Mac.

**Comment by marcodejongh** [comment]
## Publish dry-run manifests (publish_model.py, #5473)

Built and schema-validated locally with `--dry-run` (no R2 credentials in the environment); version tag `2026-09-15` (the schema's date-tag convention — the dot form from the handoff note normalized to dashes). Threshold 0.60 from the int8 tune sweep; eval block is the spray-wall `eval` split at that threshold.

### `models/hold-detector/2026-09-15/` — nano-untiled-1024
```json
{
  "schemaVersion": 1,
  "version": "2026-09-15",
  "family": "rfdetr",
  "config": "nano-untiled-1024",
  "input": {
    "width": 768,
    "height": 768,
    "layout": "NCHW",
    "dtype": "float32",
    "letterbox": "stretch",
    "normalization": {
      "mean": [
        0.485,
        0.456,
        0.406
      ],
      "std": [
        0.229,
        0.224,
        0.225
      ]
    }
  },
  "outputs": {
    "boxes": {
      "name": null,
      "format": "cxcywh-normalized"
    },
    "logits": {
      "name": null,
      "activation": "sigmoid",
      "classes": 1
    }
  },
  "thresholds": {
    "default": 0.6,
    "sweep": [
      0.05,
      0.1,
      0.15,
      0.2,
      0.3,
      0.4,
      0.5,
      0.6,
      0.7
    ]
  },
  "files": [
    {
      "path": "model-int8.onnx",
      "bytes": 31327953,
      "sha256": "84c74bc17fd078adaf33d49d8af2dc61a97daeb91b3cf0f8141cddea1c3f467c",
      "dtype": "int8"
    }
  ],
  "training": {
    "dataset": "Roboflow climbing-holds-and-volumes v14, single `hold` class, full train split (3,876 photos / 129,868 boxes)",
    "licence": "CC BY 4.0",
    "epochs": 10,
    "trainedOn": "m5-max-mps",
    "date": "2026-09-15"
  },
  "licence": "Apache-2.0",
  "eval": {
    "sprayEvalF1": 0.6588,
    "weightedCorrectionsPerHold": 0.6598,
    "gestureSavings": 0.489
  }
}
```

### `models/hold-detector/2026-09-15/` — medium-untiled-1280
```json
{
  "schemaVersion": 1,
  "version": "2026-09-15",
  "family": "rfdetr",
  "config": "medium-untiled-1280",
  "input": {
    "width": 576,
    "height": 576,
    "layout": "NCHW",
    "dtype": "float32",
    "letterbox": "stretch",
    "normalization": {
      "mean": [
        0.485,
        0.456,
        0.406
      ],
      "std": [
        0.229,
        0.224,
        0.225
      ]
    }
  },
  "outputs": {
    "boxes": {
      "name": null,
      "format": "cxcywh-normalized"
    },
    "logits": {
      "name": null,
      "activation": "sigmoid",
      "classes": 1
    }
  },
  "thresholds": {
    "default": 0.6,
    "sweep": [
      0.05,
      0.1,
      0.15,
      0.2,
      0.3,
      0.4,
      0.5,
      0.6,
      0.7
    ]
  },
  "files": [
    {
      "path": "model-int8.onnx",
      "bytes": 32846565,
      "sha256": "45df3b19400c5bed0c8d5f750c3e4b255dbdf516f1b44034dda45bc378a9455d",
      "dtype": "int8"
    }
  ],
  "training": {
    "dataset": "Roboflow climbing-holds-and-volumes v14, single `hold` class, full train split (3,876 photos / 129,868 boxes)",
    "licence": "CC BY 4.0",
    "epochs": 10,
    "trainedOn": "m5-max-mps",
    "date": "2026-09-15"
  },
  "licence": "Apache-2.0",
  "eval": {
    "sprayEvalF1": 0.6497,
    "weightedCorrectionsPerHold": 0.6577,
    "gestureSavings": 0.476
  }
}
```

Ready for the real upload with the MEDIA_* env vars set; weights upload before the manifest per the script's ordering guarantee.

**Comment by marcodejongh** [comment]
Published to the public media bucket (weights first, manifest last, per publish_model.py's ordering guarantee):

- `models/hold-detector/2026-09-15/` — **nano-untiled-1024** (the phone candidate, F1 0.659, 31.3 MB int8)
- `models/hold-detector/2026-09-15-medium/` — **medium-untiled-1280** (the server candidate, F1 0.650, 32.8 MB int8)

One layout note for #5435/#5451: the R2 prefix is one model per version (`models/hold-detector/<version>/`), so the two configs carry distinct version tags; each manifest's `config` field says which is which. The dry-run manifests in the previous comment match the uploaded ones byte for byte apart from medium's version tag. Upload verified by listing the prefix. The temporary R2 token used for this upload is being rotated.

**Comment by marcodejongh** [comment]
Implementation is now split into three draft PRs:

1. #5587: durable pg-boss recognition, dedicated CPU worker, authenticated resumable mobile flow, migration-owner queue setup, and rollout runbook. Includes the outline work from #5536 and queue foundation from #5547.
2. #5586, targeting `release/next`: removes on-device ONNX, model download/benchmark code, the iOS increased-memory entitlement, and unused JPEG dependency. Retains camera permissions.
3. [Homelab #411](https://github.com/marcodejongh/blackheathdc-ansible/pull/411): a separate Debian worker VM, digest-pinned container, private metrics/firewall, dedicated primary login and read-only R2 credential contract. Does not run on the sync VM or physical Postgres standby.

Validation so far: full typechecks, mobile suites and native bundles, real pinned-model inference, Linux image/model loading, backend authorization/retry fencing and restricted-role queue integration tests. Native simulator build/launch and 30-second crash-log check passed. Terra reviews and CI are in progress.

Not deployed or promoted: VM allocation, production credentials, published immutable image digest, reproduced offline model score, and real 24-hour/tester rollout evidence remain gates. Existing 48.2% gesture savings is prior evidence, not a new measurement. No tester counts or rollout gates have been marked complete.

The old phone-specific model-export task #5520 is retired; it no longer matches the service architecture. Track rollout evidence on #5451.
